File: examples/schema_mapping.py
"""Map schemas between two DataFrames with different column names.

GoldenFlow's SchemaMapper uses name similarity and profile similarity
to automatically map columns from a source schema to a target schema.

Usage:
    pip install goldenflow
    python examples/schema_mapping.py
"""
from __future__ import annotations

import polars as pl


if __name__ == "__main__":
    from goldenflow import SchemaMapper

    # Source has different column names than target
    source_df = pl.DataFrame({
        "fname": ["John", "Jane"],
        "lname": ["Smith", "Doe"],
        "mail": ["john@test.com", "jane@test.com"],
        "tel": ["555-1234", "555-5678"],
    })

    target_df = pl.DataFrame({
        "first_name": [""],
        "last_name": [""],
        "email": [""],
        "phone": [""],
    })

    mapper = SchemaMapper()
    mappings = mapper.map(source_df, target_df)

    print("Schema Mapping:")
    for m in mappings:
        print(f"  {m.source} -> {m.target} (confidence: {m.confidence:.0%})")

File: examples/typescript/schema-mapping.ts
/**
 * Schema mapping — auto-map columns between two datasets.
 *
 * Usage: npx tsx examples/typescript/schema-mapping.ts
 */

import { SchemaMapper } from "goldenflow";

const sourceData = [
  { fname: "John", lname: "Smith", email_address: "john@example.com", phone_number: "(555) 123-4567" },
  { fname: "Jane", lname: "Doe", email_address: "jane@test.com", phone_number: "555.987.6543" },
];

const targetSchema = [
  { first_name: "", last_name: "", email: "", phone: "", city: "" },
];

const mapper = new SchemaMapper();
const mappings = mapper.map(sourceData, targetSchema);

console.log("=== Schema Mapping ===\n");
console.log("Source Column      ->  Target Column     Confidence  Tier");
console.log("─".repeat(65));
for (const m of mappings) {
  const tier = m.confidence >= 0.9 ? "auto" : m.confidence >= 0.6 ? "suggest" : "skip";
  console.log(
    `${m.source.padEnd(18)} ->  ${m.target.padEnd(18)} ${m.confidence.toFixed(3).padEnd(11)} ${tier}`,
  );
}

// Convert to config
const config = mapper.toConfig(mappings);
console.log(`\nGenerated config with ${config.mappings.length} mapping(s)`);
